src/crudodb.ts

```diff
@@ -0,0 +1,210 @@
+         {
+           ...indexedSchema,
+           ...schema,
+           indexedIn: schema.dbVersion
+         },
+         this.idbDatabases[indexedSchema.dbName]
+       );
+       this.idbDatabases[indexedSchema.dbName] = db;
+       this.databases[key] = new Database<unknown>(
+         db.transaction(schema.store, 'readwrite')
+           .objectStore(schema.store),
+         key,
+         schema,
+         args.api
+       );
+       return key;
+    }
+
+    const itemsToUpdate = await this.general.getAll()
+      .then((items) =>
+        items.filter(({dbName}) => dbName === schema.dbName)
+      );
+    const currentVersion = evaluateDbVersion(itemsToUpdate, schema.dbName);
+    const version = currentVersion + schema.dbVersion;
+
+    const db = await prepareStoreWithDatabase({
+      ...schema,
+      dbVersion: version
+    }, this.idbDatabases[schema.dbName]);
+    this.idbDatabases[schema.dbName] = db;
+
+    const indexedEntry = await this.general.create({
+      ...schema,
+      dbVersion: version,
+      indexedIn: schema.dbVersion,
+      id: key
+    });
+    if (!indexedEntry) {
+      return;
+    }
+    this.databaseSchemas[key] = indexedEntry;
+    this.databases[key] = new Database<unknown>(
+      db.transaction(schema.store, 'readwrite')
+        .objectStore(schema.store),
+      key,
+      schema,
+      args.api
+    );
+
+    await Promise.all(
+      itemsToUpdate.map((item) => ({
+        ...item,
+        dbVersion: version
+      })).map((item) => this.general.update(item))
+    );
+
+    return key;
+  }
+
+  private async setup(): Promise<void> {
+    const start = +new Date();
+    const { db, database } = await initGeneralDb();
+    this.databases.general = database as Database<unknown>;
+    this.idbDatabases['__GENERAL__'] = db;
+    if (this.debug) {
+      console.debug('synced databases', {
+        took: +new Date() - start
+      });
+    }
+  }
+
+  public async sync(schemaKey?: string[]): Promise<void> {
+    const toUpdate = schemaKey ? Object.entries(this.databases).filter(([key, _]) => schemaKey.includes(key)) : Object.entries(this.databases);
+    await Promise.all(
+      toUpdate.map(([key, db]) => db.sync().catch((err) =>
+        console.warn(`db with key ${key} failed to sync`, err)
+      ))
+    );
+  }
+}
+
+async function upgradeExistingSchema(
+  general: Database<InternalStoreEntry>,
+  indexedSchema: InternalStoreEntry,
+  openDatabase?: IDBDatabase
+): Promise<IDBDatabase> {
+  const internals = await general.getAll()
+    .then((databases) =>
+      databases.filter(({ dbName }) => dbName === indexedSchema.dbName)
+    );
+  const version = evaluateDbVersion(internals, indexedSchema.dbName);
+  const withNewVersion = internals.map((item) => ({
+    ...item,
+    dbVersion: version
+  }));
+  const db = await prepareStoreWithDatabase(indexedSchema, openDatabase);
+  await Promise.all(withNewVersion.map((item) => general.update(item)));
+  await general.update(indexedSchema);
```

Comment:
It's not immediately clear if the above three lines are possible to run in parallel or not. I suggest to add comments if they depend on each other to outline their dependency.